token-swap/program/src/curve/stable.rs

```diff
@@ -0,0 +1,467 @@
+    pub owner_trade_fee_numerator: u64,
+    /// Owner trade fee denominator
+    pub owner_trade_fee_denominator: u64,
+    /// Owner withdraw fee numerator
+    pub owner_withdraw_fee_numerator: u64,
+    /// Owner withdraw fee denominator
+    pub owner_withdraw_fee_denominator: u64,
+    /// Host trading fee numerator
+    pub host_fee_numerator: u64,
+    /// Host trading fee denominator
+    pub host_fee_denominator: u64,
+    /// Amplifier constant
+    pub amp: u64,
+}
+
+fn calc_step(
+    initial_d: u128,
+    leverage: u128,
+    sum_x: u128,
+    d_p: u128,
+    n_coins: u128,
+) -> Option<u128> {
+    let leverage_mul = leverage.checked_mul(sum_x)?;
+    let d_p_mul = d_p.checked_mul(n_coins)?;
+
+    let l_val = leverage_mul.checked_add(d_p_mul)?.checked_mul(initial_d)?;
+
+    let leverage_sub = leverage.checked_sub(1)?.checked_mul(initial_d)?;
+    let n_coins_sum = n_coins.checked_add(1)?.checked_mul(d_p)?;
+
+    let r_val = leverage_sub.checked_add(n_coins_sum)?;
+
+    l_val.checked_div(r_val)
+}
+
+/// Compute stable swap invariant (D)
+/// Equation:
+/// A * sum(x_i) * n**n + D = A * D * n**n + D**(n+1) / (n**n * prod(x_i))
+fn compute_d(amp: u128, amount_a: u128, amount_b: u128) -> Option<u128> {
+    // XXX: Curve uses u256
+    let n_coins: u128 = 2; // n
+    let amount_a_times_coins = amount_a.checked_mul(n_coins)?;
+    let amount_b_times_coins = amount_b.checked_mul(n_coins)?;
+    let sum_x = amount_a.checked_add(amount_b)?; // sum(x_i), a.k.a S
+    if sum_x == 0 {
+        Some(0)
+    } else {
+        let mut d_prev: u128;
+        let mut d = sum_x;
+        let leverage = amp.checked_mul(n_coins)?; // A * n
+                                                  // Newton's method to approximate D
+        for _ in 0..128 {
+            let mut d_p = d;
+            d_p = d_p.checked_mul(d)?.checked_div(amount_a_times_coins)?;
+            d_p = d_p.checked_mul(d)?.checked_div(amount_b_times_coins)?;
+            d_prev = d;
+            //d = (leverage * sum_x + d_p * n_coins) * d / ((leverage - 1) * d + (n_coins + 1) * d_p);
+            d = calc_step(d, leverage, sum_x, d_p, n_coins)?;
+            // Equality with the precision of 1
+            if d > d_p {
+                if d.checked_sub(d_prev)? <= 1 {
+                    break;
+                }
+            } else if d_prev.checked_sub(d)? <= 1 {
+                break;
+            }
+        }
+        Some(d)
+    }
+}
+
+/// Compute swap amount `y` in proportion to `x`
+/// Solve for y:
+/// y**2 + y * (sum' - (A*n**n - 1) * D / (A * n**n)) = D ** (n + 1) / (n ** (2 * n) * prod' * A)
+/// y**2 + b*y = c
+fn compute_dest(amp: u128, new_source_amount: u128, d_val: u128) -> Option<u128> {
+    // XXX: Curve uses u256
+    let n_coins: u128 = 2;
+    let leverage = amp.checked_mul(n_coins)?; // A * n
+
+    // sum' = prod' = x
+    // c =  D ** (n + 1) / (n ** (2 * n) * prod' * A)
+    let c = checked_pow(d_val, 3)?.checked_div(
```

Comment:
better to use `checked_pow(d_val, n_coins.checked_add(1)?)` to avoid an issue later